### PR TITLE
enable privateuseone to perform streaming backward

### LINKED ISSUE #116957: enable privateuseone to perform multiple streams backwards
### 🚀 The feature, motivation and pitch

My team is working on adding custom backend for pytorch, and would like multiple streams backwards to be possible on privateuseone backend.  
Specifically, part of the code in the autograd engine was hardcoded into CUDA without considering other accelerators similar to CUDA, resulting in other custom backends being unable to use multi-stream training.
For example, 
https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/engine.cpp#L1524  
```
void GraphTask::stash_current_streams() {
  const auto guard = c10::impl::VirtualGuardImpl{c10::DeviceType::CUDA};
  auto num_gpus = guard.deviceCount();
  caller_current_streams_.resize(num_gpus);
  if (num_gpus > 0) {
    for (c10::DeviceIndex idx = 0; idx < num_gpus; idx++) { 
```  
https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/engine.cpp#L716  
``` 
if (!leaf_streams.empty()) {
    for (const auto& leaf_stream : leaf_streams) {
      // stash_current_streams() stashed streams for all device IDs that already
      // had a CUDA context before the GraphTask executed. For inactive devices,
      // it stashed a c10::nullopt. I don't expect GraphTask's backward pass ran
      // leaf nodes on any new devices, so the stashed streams should be enough.
      // If leaf_stream.device_index() happens to be for a new device,
      // operator* on the c10::nullopt should throw an error.
      const auto caller_current_stream =
          // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
          *caller_current_streams_[leaf_stream.device_index()];

      if (caller_current_stream != leaf_stream) {
        auto event = c10::Event{c10::DeviceType::CUDA};
```
Of course, the above two code snippets are just for illustration.

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @bdhirsh

### PR DESCRIPTION
**Pull request by garfield1997**
Fixes #116957

cc @albanD @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Change sounds good!
Small nit to add an extra check to ensure the assumption is valid

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -1160,6 +1179,11 @@ auto Engine::compute_dependencies(
     // so GraphTask::exec_post_processing can sync them with leaf_streams.
     task.stash_current_streams();
   }
+
+  if (will_use_privateuse1) {
```
[the remark this answers is not in the document]
Comment (garfield1997) [reply torch/csrc/autograd/engine.cpp:1188]: Ok, I just added it.

**Review by albanD — COMMENTED** [review]
Sounds good to me, will let @soulitzer do the final review.

@garfield1997 Can you please sign the CLA otherwise we won't be able to merge the PR.

**Comment by garfield1997** [comment]
I just signed the CLA. This is my first time contributing code, and it took quite a while. I apologize for the delay. @albanD

**Comment by garfield1997** [comment]
@soulitzer If you have time, could you please help review this commit?

**Comment by garfield1997** [comment]
I just rebased this commit. If you have any further questions, feel free to bring them up. @albanD  @soulitzer

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -1545,6 +1578,24 @@ void GraphTask::stash_current_streams() {
   }
 }
 
+// Remembers current streams on all devices where a context has been created for
+// privateuse1
+void GraphTask::stash_current_privateuse1_streams() {
```
Comment (soulitzer) [inline torch/csrc/autograd/engine.cpp:1583]: nit: can we rename the existing `stash_current_streams` to `stash_current_cuda_streams` 
(don't forget to update the comment above it as well)

File: torch/csrc/autograd/input_buffer.cpp [hunk 3]
```diff
@@ -197,10 +198,10 @@ void InputBuffer::add(
       if (opt_sync_stream && (opt_accumulate_stream != opt_sync_stream)) {
         // (3b), (4b)
         c10::OptionalDeviceGuard device_guard{opt_sync_stream->device()};
-        auto event = c10::Event{c10::DeviceType::CUDA};
+        auto event = c10::Event{device_type};
         event.record(*opt_sync_stream);
         opt_accumulate_stream->wait(event);
-        const auto guard = c10::impl::VirtualGuardImpl(c10::DeviceType::CUDA);
+        const auto guard = c10::impl::VirtualGuardImpl(device_type);
```
Comment (soulitzer) [inline torch/csrc/autograd/input_buffer.cpp:204]: Update comment below:
```
// (1) non-CUDA variable
```

File: torch/csrc/autograd/input_buffer.cpp [hunk 4]
```diff
@@ -160,8 +160,9 @@ void InputBuffer::add(
 
   TORCH_INTERNAL_ASSERT(device_of(var));
   c10::optional<c10::Stream> opt_accumulate_stream = c10::nullopt;
+  const auto device_type = device_of(var).value().type();
```
Comment (soulitzer) [inline torch/csrc/autograd/input_buffer.cpp:163]: Update comment above

File: torch/csrc/autograd/engine.cpp [hunk 5]
```diff
@@ -713,7 +714,7 @@ void GraphTask::exec_post_processing() {
           *caller_current_streams_[leaf_stream.device_index()];
 
       if (caller_current_stream != leaf_stream) {
-        auto event = c10::Event{c10::DeviceType::CUDA};
+        auto event = c10::Event{leaf_stream.device_type()};
```
Comment (soulitzer) [inline torch/csrc/autograd/engine.cpp:717]: Update comment above

Comment (soulitzer) [reply torch/csrc/autograd/engine.cpp:717]: Also `Note [Streaming backwards]`

**Review by soulitzer — COMMENTED** [review]
Thanks, sorry for the delay! This looks good to me, but could you double check for any other comments that should be updated?

File: torch/csrc/autograd/engine.cpp (hunk 2 above)
Comment (garfield1997) [reply torch/csrc/autograd/engine.cpp:1583]: Ok，I changed the name and added privateuse1 to the comments.

**Comment by garfield1997** [comment]
I ran lint locally, sorry for not doing this at the beginning.

**Comment by garfield1997** [comment]
It seems that the CI failure is unrelated to the changes. I ran this test locally and was unable to reproduce the issue.

**Review by soulitzer — APPROVED** [review]
Thanks for checking, this LGTM

**Pull request closed without merging**